utils/transfer/src/progress.rs

```diff
@@ -0,0 +1,93 @@
+use crate::error::Error;
+use crate::{abortable_sink, abortable_stream, TransferSink, TransferStream};
+use crate::{TransferContext, TransferData};
+use futures::{SinkExt, StreamExt, TryFutureExt};
+use tokio::task::spawn_local;
+
+type Stream = TransferStream<TransferData, Error>;
+
+/// Wraps a stream to report progress.
+/// The `report` function is called with the current offset and the total size.
+/// The total size is 0 if the size is unknown. (For example, when the source is a directory.)
+/// The reporting function should not block as it will block the transfer.
```

Comment:
I would recommend using `Option::None` in case size is unknown. This way we don't have magic number with special semantic.
Other possibility is to create enum with variant Unknown